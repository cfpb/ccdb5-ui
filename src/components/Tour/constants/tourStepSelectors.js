export const DOCUMENT_STEP_SELECTORS = {
  STEP_1: '.content__hero',
  STEP_2: '.m-pagination__btn-next',
  STEP_3: '.saved__search-panel div:nth-child(2)',
  STEP_4: '.back-to-results',
};

export const GEO_STEP_SELECTORS = {
  STEP_1: { selector: '.content', label: 'Index' },
  STEP_2: {
    selector: 'header.content__hero .m-list--horizontal',
    label: 'Help Links',
  },
  STEP_3: { selector: '.tabbed-navigation', label: 'Views' },
  STEP_4: { selector: '.filter-panel', label: 'Filters' },
  STEP_5: { selector: '.date-filter', label: 'Date Received' },
  STEP_6: {
    selector: '.filter-panel .aggregation.product',
    label: 'Product/Issue',
  },
  STEP_7: {
    selector: '.filter-panel .aggregation button',
    label: 'Show/Hide',
  }, // show/hide filters
  STEP_8: { selector: '.search-bar', label: 'Search Bar' },
  STEP_9: { selector: '#search-summary h2', label: 'Search Summary' },
  STEP_10: {
    selector: '.export-results button.export-btn',
    label: 'Export Data',
  },
  STEP_11: {
    selector: '.export-results button.print-preview',
    label: 'Print Page',
  }, // print button
  STEP_12: { selector: '.map-shading', label: 'Map Shading' },
  STEP_13: { selector: '#tile-chart-map', label: 'Tile Map' },
  STEP_14: { selector: '.row-chart-section', label: 'Row Charts' },
  STEP_15: { selector: '.tabbed-navigation', label: 'Tabbed Navigation' },
};

export const LIST_COMPLAINTS_SELECTORS = {
  STEP_1: { selector: '.content', label: 'Index' },
  STEP_2: {
    selector: 'header.content__hero .m-list--horizontal',
    label: 'Help Links',
  },
  STEP_3: { selector: '.tabbed-navigation', label: 'Views' },
  STEP_4: { selector: '.filter-panel', label: 'Filters' },
  STEP_5: { selector: '.date-filter', label: 'Date Received' },
  STEP_6: {
    selector: '.filter-panel .aggregation.product',
    label: 'Product/Issue',
  },
  STEP_7: {
    selector: '.filter-panel .aggregation button',
    label: 'Show/Hide',
  }, // show/hide filters
  STEP_8: { selector: '.search-bar', label: 'Search Bar' },
  STEP_9: { selector: '#search-summary h2', label: 'Search Summary' },
  STEP_10: {
    selector: '.export-results button.export-btn',
    label: 'Export Data',
  },
  STEP_11: {
    selector: '.export-results button.print-preview',
    label: 'Print Page',
  }, // print button
  STEP_12: { selector: '[data-tour="select-size"]', label: 'Show Number' },
  STEP_13: { selector: '[data-tour="select-sort"]', label: 'Sort Results' },
  STEP_14: {
    selector: '.refine-bar section.m-btn-group',
    label: 'Show Narratives',
  }, // read complaints
  STEP_15: { selector: '.cards-panel', label: 'Results' },
  STEP_16: { selector: '.to-detail', label: 'View Detailed Complaints' }, // complaint id
  STEP_17: { selector: '.tabbed-navigation', label: 'Tabbed Navigation' },
};

export const TRENDS_SELECTORS = {
  STEP_1: { selector: '.content', label: 'Index' },
  STEP_2: {
    selector: 'header.content__hero .m-list--horizontal',
    label: 'Help Links',
  },
  STEP_3: { selector: '.tabbed-navigation', label: 'Views' },
  STEP_4: { selector: '.filter-panel', label: 'Filters' },
  STEP_5: { selector: '.date-filter', label: 'Date Received' },
  STEP_6: {
    selector: '.filter-panel .aggregation.product',
    label: 'Product/Issue',
  },
  STEP_7: {
    selector: '.filter-panel .aggregation button',
    label: 'Show/Hide',
  }, // show/hide filters
  STEP_8: { selector: '.search-bar', label: 'Search Bar' },
  STEP_9: { selector: '#search-summary h2', label: 'Search Summary' },
  STEP_10: {
    selector: '.export-results button.export-btn',
    label: 'Export Data',
  },
  STEP_11: {
    selector: '.export-results button.print-preview',
    label: 'Print Page',
  }, // print button
  STEP_12: { selector: '[data-tour="select-lens"]', label: 'Aggregate By' },
  STEP_13: {
    selector: '[data-tour="select-interval"]',
    label: 'Date Interval',
  },
  STEP_14: { selector: '.row-chart-section', label: 'Row Charts' },
  STEP_15: { selector: '.tabbed-navigation', label: 'Tabbed Navigation' },
};
