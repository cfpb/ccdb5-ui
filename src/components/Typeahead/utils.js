/**
 * Sets the options for the async typeahead
 *
 * @param {string} value - value typed into the typeahead
 * @param {Function} setOptions - function to set local array of options
 * @param {string} uri - the endpoint being hit by the fetch
 * @returns {void | Promise<object>} - Either nothing or a response with option object
 */
export function handleFetchSearch(value, setOptions, uri) {
  const num = value.toLowerCase();
  if (num === '') {
    setOptions([]);
    return;
  }
  return fetch(uri)
    .then((result) => result.json())
    .then((items) => {
      const options = items.map((item) => ({
        key: item,
        label: item,
        position: item.toLowerCase().indexOf(num),
        value,
      }));
      setOptions(options);
    });
}
