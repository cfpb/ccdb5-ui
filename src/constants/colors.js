module.exports = Object.freeze({
  BriteCharts: {
    dark: '#118935',
    regular: '#20aa3f',
    medium: '#ADDC91',
    light: '#C7E5B3',
  },
  // if you modify anything here,
  // make sure that the following colors are in sync
  // Line Chart
  // Area Chart
  // Stacked Area Chart
  // External Tooltip/Legend
  DataLens: {
    0: '#2cb34a', // 0
    1: '#addc91', // 1
    2: '#0072ce', // 4
    3: '#96c4ed', // 5
    4: '#257675', // 2
    5: '#9ec4c3', // 3
    6: '#254b87',
    7: '#9daecc',
    8: '#b4267a',
    9: '#dc9cbf',
    10: '#a2a3a4',
  },
});
