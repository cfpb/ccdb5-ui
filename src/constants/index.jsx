// Internal triggers
export const API_CALLED = 'API_CALLED'
export const MODAL_HID = 'MODAL_HID'
export const MODAL_SHOWN = 'MODAL_SHOWN'

// Modal Dialogs
export const MODAL_TYPE_DATA_EXPORT = 'MODAL_TYPE_DATA_EXPORT'
export const MODAL_TYPE_MORE_ABOUT = 'MODAL_TYPE_MORE_ABOUT'

// Useful constants
export const NARRATIVE_SEARCH_FIELD = 'complaint_what_happened'

// query manager flags
// These constants control how the query manager works
export const REQUERY_ALWAYS = 'REQUERY_ALWAYS'
export const REQUERY_HITS_ONLY = 'REQUERY_HITS_ONLY'
// default if not specified
export const REQUERY_NEVER = 'REQUERY_NEVER'

export const knownFilters = [
  'company',
  'company_public_response',
  'company_response',
  'consumer_consent_provided',
  'consumer_disputed',
  'issue',
  'product',
  'state',
  'submitted_via',
  'tag',
  'timely',
  'zip_code'
]

export const dateFilters = [
  'company_received_max',
  'company_received_min',
  'date_received_max',
  'date_received_min'
]

export const flagFilters = [
  'has_narrative'
]

export const SLUG_SEPARATOR = 'â€¢'

// Keyboard
export const VK_DOWN = 40
export const VK_ENTER = 14
export const VK_ESCAPE = 27
export const VK_RETURN = 13
export const VK_TAB = 9
export const VK_UP = 38

/// ////////////////////////////////////////////////////////////////////////////
// JSON Objects

export const THESE_UNITED_STATES = {
  AA: 'Armed Forces Americas',
  AE: 'Armed Forces Europe',
  AL: 'Alabama',
  AK: 'Alaska',
  AS: 'American Samoa',
  AP: 'Armed Forces Pacific',
  AZ: 'Arizona',
  AR: 'Arkansas',
  CA: 'California',
  CO: 'Colorado',
  CT: 'Connecticut',
  DE: 'Delaware',
  DC: 'District Of Columbia',
  FM: 'Federated States Of Micronesia',
  FL: 'Florida',
  GA: 'Georgia',
  GU: 'Guam',
  HI: 'Hawaii',
  ID: 'Idaho',
  IL: 'Illinois',
  IN: 'Indiana',
  IA: 'Iowa',
  KS: 'Kansas',
  KY: 'Kentucky',
  LA: 'Louisiana',
  ME: 'Maine',
  MH: 'Marshall Islands',
  MD: 'Maryland',
  MA: 'Massachusetts',
  MI: 'Michigan',
  MN: 'Minnesota',
  MS: 'Mississippi',
  MO: 'Missouri',
  MT: 'Montana',
  NE: 'Nebraska',
  NV: 'Nevada',
  NH: 'New Hampshire',
  NJ: 'New Jersey',
  NM: 'New Mexico',
  NY: 'New York',
  NC: 'North Carolina',
  ND: 'North Dakota',
  MP: 'Northern Mariana Islands',
  OH: 'Ohio',
  OK: 'Oklahoma',
  OR: 'Oregon',
  PW: 'Palau',
  PA: 'Pennsylvania',
  PR: 'Puerto Rico',
  RI: 'Rhode Island',
  SC: 'South Carolina',
  SD: 'South Dakota',
  TN: 'Tennessee',
  TX: 'Texas',
  UT: 'Utah',
  VT: 'Vermont',
  VI: 'Virgin Islands',
  VA: 'Virginia',
  WA: 'Washington',
  WV: 'West Virginia',
  WI: 'Wisconsin',
  WY: 'Wyoming'
}

export const TILE_MAP_STATES = [
  'AK',
  'AL',
  'AR',
  'AZ',
  'CA',
  'CO',
  'CT',
  'DC',
  'DE',
  'FL',
  'GA',
  'HI',
  'IA',
  'ID',
  'IL',
  'IN',
  'KS',
  'KY',
  'LA',
  'MA',
  'MD',
  'ME',
  'MI',
  'MN',
  'MO',
  'MS',
  'MT',
  'NC',
  'ND',
  'NE',
  'NH',
  'NJ',
  'NM',
  'NV',
  'NY',
  'OH',
  'OK',
  'OR',
  'PA',
  'RI',
  'SC',
  'SD',
  'TN',
  'TX',
  'UT',
  'VA',
  'VT',
  'WA',
  'WI',
  'WV',
  'WY'
]
