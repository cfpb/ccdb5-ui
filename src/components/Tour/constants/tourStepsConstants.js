import { MODE_LIST, MODE_MAP, MODE_TRENDS } from '../../../constants';

import {
  GEO_STEP_SELECTORS as GEO,
  LIST_COMPLAINTS_SELECTORS as LIST,
  TRENDS_SELECTORS as TRENDS,
} from './tourStepSelectors';

const generateLinkText = (index, linkObject) =>
  `<li><button class="a-btn a-btn--link" onclick="document.querySelectorAll('.introjs-bullets li a')[${
    index + 1
  }].click()">` +
  linkObject[1].label +
  '</button></li>';

let geoIndex = '';

// we exclude the first entry so we can have Index appear as the title
Object.entries(GEO)
  .slice(1)
  .forEach((value, key) => {
    geoIndex += generateLinkText(key, value);
  });

let listIndex = '';
Object.entries(LIST)
  .slice(1)
  .forEach((value, key) => {
    listIndex += generateLinkText(key, value);
  });

let trendsIndex = '';
Object.entries(TRENDS)
  .slice(1)
  .forEach((value, key) => {
    trendsIndex += generateLinkText(key, value);
  });

export const TOUR_STEPS = {
  [MODE_MAP]: [
    {
      element: GEO.STEP_1.selector,
      intro:
        '<div>' +
        '<div class="left">' +
        'Welcome to the Consumer Complaint Database!' +
        '<br /><br />' +
        'Complaints the CFPB sends to companies for response are published in the Consumer Complaint Database after the company responds, confirming a commercial relationship with the consumer, or after 15 days, whichever comes first.' +
        '<br /><br />' +
        'Complaints the CFPB refers to other regulators, such as complaints about depository institutions with less than $10 billion in assets, are not published in the database. This database is not a statistical sample of consumers’ experiences in the marketplace.' +
        '</div>' +
        '<div class="right">' +
        '<h4>Index</h4>' +
        '<ul>' +
        geoIndex +
        '</ul></div>' +
        '</div>',
      tooltipClass: 'wide',
    },
    {
      element: GEO.STEP_2.selector,
      intro: 'These links provide more information about this database.',
    },
    {
      element: GEO.STEP_3.selector,
      intro:
        'You can change how to view complaint data by selecting the three views we have available — Trends, List, and Map.' +
        '<br /><br />' +
        'You are currently on the Map view, which displays complaints by state.' +
        '<br /><br />' +
        'To read individual complaints, select List view' +
        '<br /><br />' +
        'To visualize complaint data with product and issue breakdowns, select Trends view',
      position: 'top',
    },
    {
      element: GEO.STEP_4.selector,
      intro:
        'The “Filter results by” panel lets you filter complaint data. To clear filters, uncheck individual selections in the filter panel or to select “Clear all filters” under the Search bar.',
    },
    {
      element: GEO.STEP_5.selector,
      intro:
        'You can adjust the date range of the complaint data showed using the “Date CFPB received the complaint” filter. <br /> The default date range is three years from today’s date.',
    },
    {
      element: GEO.STEP_6.selector,
      intro:
        'You can filter complaints by the product and issue the consumer selected when they submitted their complaint.' +
        '<br /><br />' +
        'Expand a product or issue to see its sub-products or sub-issues, where available. ',
    },
    {
      element: GEO.STEP_7.selector,
      intro:
        'The blue show/hide icon allows you to expand or collapse specific filters.',
    },
    {
      element: GEO.STEP_8.selector,
      intro:
        'Use the “Search within” bar to find specific words or phrases in complaints. By default, this will search “All data”, but you can change this using the gray dropdown menu.' +
        '<br /> <br />' +
        'Click “Show advanced search tips” to see some advanced ways to refine your word search.',
    },
    {
      element: GEO.STEP_9.selector,
      intro:
        'As you apply filters and search terms, this will display how many complaints are included out of the total number of complaints published in the database.',
    },
    {
      element: GEO.STEP_10.selector,
      intro:
        'You can export full or filtered results to CSV or JSON by selecting “Export data”.',
    },
    {
      element: GEO.STEP_11.selector,
      intro: 'Select “Print” to generate a pdf of the current page.',
    },
    {
      element: GEO.STEP_12.selector,
      intro:
        'Map shading defaults to total complaints, but you can change this to show complaints per 1,000 population in that state.',
    },
    {
      element: GEO.STEP_13.selector,
      intro:
        'Hover over a tile on the map to view complaint information specific to that state. You can filter to see data specific to a state by selecting the state or using the filter panel.',
    },
    {
      element: GEO.STEP_14.selector,
      intro:
        'View complaint volume by product in the bar chart. Use the blue arrow by each product to view breakdowns by sub-products.' +
        '<br /><br />Once the product is expanded, select the “Visualize” link as shown below, to view trends on sub-products and issues:',
    },
    {
      element: GEO.STEP_15.selector,
      intro:
        'To learn how to use Trends and List views, select the view and then start the tour.',
    },
  ],
  [MODE_LIST]: [
    {
      element: LIST.STEP_1.selector,
      intro:
        '<div>' +
        '<div class="left">' +
        'Welcome to the Consumer Complaint Database!' +
        '<br /><br />' +
        'Complaints the CFPB sends to companies for response are published in the Consumer Complaint Database after the company responds, confirming a commercial relationship with the consumer, or after 15 days, whichever comes first.' +
        '<br /><br />' +
        'Complaints the CFPB refers to other regulators, such as complaints about depository institutions with less than $10 billion in assets, are not published in the database. This database is not a statistical sample of consumers’ experiences in the marketplace.' +
        '</div>' +
        '<div class="right">' +
        '<h4>Index</h4>' +
        '<ul>' +
        listIndex +
        '</ul></div>' +
        '</div>',
      tooltipClass: 'wide',
    },
    {
      element: LIST.STEP_2.selector,
      intro: 'These links provide more information about this database.',
    },
    {
      element: LIST.STEP_3.selector,
      intro:
        'You can change how to view complaint data by selecting the three views we have available — Trends, List, and Map.' +
        '<br /><br />' +
        'You are currently on the List view, which shows individual complaints.' +
        '<br /><br />' +
        'To visualize complaint data with product and issue breakdowns, select Trends view.' +
        '<br /><br />' +
        'To view complaints by state, select Map view.',
    },
    {
      element: LIST.STEP_4.selector,
      intro:
        'The “Filter results by” panel lets you filter complaint data. To clear filters, uncheck individual selections in the filter panel or to select “Clear all filters” under the Search bar.',
    },
    {
      element: LIST.STEP_5.selector,
      intro:
        'You can adjust the date range of the complaint data showed using the “Date CFPB received the complaint” filter. <br /> The default date range is three years from today’s date.',
    },
    {
      element: LIST.STEP_6.selector,
      intro:
        'You can filter complaints by the product and issue the consumer selected when they submitted their complaint.' +
        '<br /><br />' +
        'Expand a product or issue to see its sub-products or sub-issues, where available. ',
    },
    {
      element: LIST.STEP_7.selector,
      intro:
        'The blue show/hide icon allows you to expand or collapse specific filters.',
    },
    {
      element: LIST.STEP_8.selector,
      intro:
        'Use the “Search within” bar to find specific words or phrases in complaints. By default, this will search “All data”, but you can change this using the gray dropdown menu.' +
        '<br /> <br />' +
        'Click “Show advanced search tips” to see some advanced ways to refine your word search.',
    },
    {
      element: LIST.STEP_9.selector,
      intro:
        'As you apply filters and search terms, this will display how many complaints are included out of the total number of complaints published in the database.',
    },
    {
      element: LIST.STEP_10.selector,
      intro:
        'You can export full or filtered results to CSV or JSON by selecting “Export data”.',
    },
    {
      element: LIST.STEP_11.selector,
      intro: 'Select “Print” to generate a pdf of the current page.',
    },
    {
      element: LIST.STEP_12.selector,
      intro: 'Select the number of complaints to view per page here.',
    },
    {
      element: LIST.STEP_13.selector,
      intro:
        'Sort complaints in the list view by “Newest to oldest,” “Oldest to newest,” “Relevance,” and “Relevance (asc).”' +
        '<br /><br />' +
        'Relevancy are based on any applied search terms.',
    },
    {
      element: LIST.STEP_14.selector,
      intro:
        'You can see all complaints or focus on those with consumer complaint narratives.' +
        '<br /><br />' +
        'We publish the consumer’s narrative description of what happened from their complaint if the consumer opts to share it publicly and after taking steps to remove personal information.',
    },
    {
      element: LIST.STEP_15.selector,
      intro:
        'These are your search results. As you apply filters and search terms, your results will display here. This view shows a preview of the complaint.',
    },
    {
      element: LIST.STEP_16.selector,
      intro:
        'Click the blue ID number to see more details about a particular complaint.',
    },
    {
      element: LIST.STEP_17.selector,
      intro:
        'To learn how to use Map and Trend views, select the view and then start the tour.',
    },
  ],
  [MODE_TRENDS]: [
    {
      element: TRENDS.STEP_1.selector,
      intro:
        '<div>' +
        '<div class="left">' +
        'Welcome to the Consumer Complaint Database!' +
        '<br /><br />' +
        'Complaints the CFPB sends to companies for response are published in the Consumer Complaint Database after the company responds, confirming a commercial relationship with the consumer, or after 15 days, whichever comes first.' +
        '<br /><br />' +
        'Complaints the CFPB refers to other regulators, such as complaints about depository institutions with less than $10 billion in assets, are not published in the database. This database is not a statistical sample of consumers’ experiences in the marketplace.' +
        '</div>' +
        '<div class="right">' +
        '<h4>Index</h4>' +
        '<ul>' +
        trendsIndex +
        '</ul></div>' +
        '</div>',
      tooltipClass: 'wide first-step',
    },
    {
      element: TRENDS.STEP_2.selector,
      intro: 'These links provide more information about this database.',
    },
    {
      element: TRENDS.STEP_3.selector,
      intro:
        'You can change how to view complaint data by selecting the three views we have available—Trends, List, and Map.' +
        '<br /><br />' +
        'You are currently on the Trends view, which visualizes complaint data using a trend line with product and issue breakdowns.' +
        '<br /><br />' +
        'To read individual complaints, select List view.' +
        '<br /><br />' +
        'To view complaints by state, select Map view.',
      position: 'top',
    },
    {
      element: TRENDS.STEP_4.selector,
      intro:
        'The “Filter results by” panel lets you filter complaint data. To clear filters, uncheck individual selections in the filter panel or to select “Clear all filters” under the Search bar.',
    },
    {
      element: TRENDS.STEP_5.selector,
      intro:
        'You can adjust the date range of the complaint data showed using the “Date CFPB received the complaint” filter. <br /> The default date range is three years from today’s date.',
    },
    {
      element: TRENDS.STEP_6.selector,
      intro:
        'You can filter complaints by the product and issue the consumer selected when they submitted their complaint.' +
        '<br /><br />' +
        'Expand a product or issue to see its sub-products or sub-issues, where available. ',
    },
    {
      element: TRENDS.STEP_7.selector,
      intro:
        'The blue show/hide icon allows you to expand or collapse specific filters.',
    },
    {
      element: TRENDS.STEP_8.selector,
      intro:
        'Use the “Search within” bar to find specific words or phrases in complaints. By default, this will search “All data”, but you can change this using the gray dropdown menu.' +
        '<br /> <br />' +
        'Click “Show advanced search tips” to see some advanced ways to refine your word search.',
    },
    {
      element: TRENDS.STEP_9.selector,
      intro:
        'As you apply filters and search terms, this will display how many complaints are included out of the total number of complaints published in the database.',
    },
    {
      element: TRENDS.STEP_10.selector,
      intro:
        'You can export full or filtered results to CSV or JSON by selecting “Export data”.',
    },
    {
      element: TRENDS.STEP_11.selector,
      intro: 'Select “Print” to generate a pdf of the current page.',
    },
    {
      element: TRENDS.STEP_12.selector,
      intro:
        'Select the dropdown to aggregate the data by product or by the company to which the CFPB sent the complaint for response.',
    },
    {
      element: TRENDS.STEP_13.selector,
      intro:
        'Select this dropdown to change the time interval displayed on the trend line.',
    },
    {
      element: TRENDS.STEP_14.selector,
      intro:
        'View complaint volume by product in the bar chart. Use the blue arrow by each product to view breakdowns by sub-products.' +
        '<br /><br />Once the product is expanded, select the “Visualize” link as shown below, to view trends on sub-products and issues:',
    },
    {
      element: TRENDS.STEP_15.selector,
      intro:
        'To learn how to use List and Map views, select the view and then start the tour.',
      position: 'top',
    },
  ],
};
