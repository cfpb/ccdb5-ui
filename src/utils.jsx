import { DATE_RANGE_MIN, SLUG_SEPARATOR } from './constants'
import moment from 'moment'

// eslint-disable-next-line complexity
export const calculateDateInterval = ( minDate, maxDate ) => {
  // only check intervals if the end date is today
  // round off the date so the partial times don't mess up calculations
  const today = moment( new Date() ).startOf( 'day' )
  const end = moment( maxDate ).startOf( 'day' )
  const start = moment( minDate ).startOf( 'day' )

  // make sure end date is the same as today's date
  if ( end.diff( today, 'days', true ) !== 0 ) {
    return ''
  }

  // is the start date the same as the oldest document?
  if ( moment( minDate ).isSame( DATE_RANGE_MIN, 'day' ) ) {
    return 'All'
  }

  // verify if it's 3 or 1 years
  const yrDiff = end.diff( start, 'years', true )

  if ( yrDiff === 3 || yrDiff === 1 ) {
    return yrDiff + 'y'
  }

  // verify if it's 6 or 3 months
  const moDiff = end.diff( start, 'months', true )
  if ( moDiff === 6 || moDiff === 3 ) {
    return moDiff + 'm'
  }

  return ''
}

/**
 * Function to set the limit of the range of a set of numbers
 * @param {int} x value we are checking
 * @param {int} min smallest number it can me
 * @param {int} max biggest number it can be
 * @returns {*}the limited value
 */
export const clamp = ( x, min, max ) => {
  if ( x < min ) {
    x = min;
  } else if ( x > max ) {
    x = max;
  }
  return x;
}

/**
 * Replacement for the common pattern:
 * if( o.field )
 *    x = o.field
 * else
 *    x = alternateValue
 *
 * Avoids some of the complexity lint warnings
 *
 * @param {Object} o the object being tested
 * @param {string} field the field to check
 * @param {string|Object} alternateValue the value to use in absence
 * @returns {string} the value to use
 */
export const coalesce = ( o, field, alternateValue ) => {
  if ( typeof o !== 'object' ) {
    return alternateValue;
  }

  return field in o && o[field] ? o[field] : alternateValue;
};

/**
 * Creates a hash from a string
 *
 * @param {string} someString the string to hash
 * @returns {int} a hashing of the string
 */
export function hashCode( someString ) {
  const s = String( someString )
  let hash = 0
  let i, chr
  if ( s.length === 0 ) {
    return hash
  }
  for ( i = 0; i < s.length; i++ ) {
    chr = s.charCodeAt( i )
    hash = ( hash << 5 ) - hash + chr

    // Convert to 32bit integer
    hash |= 0
  }
  return hash
}

/**
 * Creates a hash from an object
 *
 * @param {string} o the object to hash
 * @returns {string} a hashing of the object
 */
export function hashObject( o ) {
  return hashCode( JSON.stringify( o ) )
}
export const normalize = s => s.toLowerCase()

export const slugify = ( a, b ) => a + SLUG_SEPARATOR + b

export const sortSelThenCount = ( options, selected ) => {
  const retVal = ( options || [] ).slice()

  // Sort the array so that selected items appear first, then by doc_count
  /* eslint complexity: ["error", 5] */
  retVal.sort( ( a, b ) => {
    const aSel = selected.indexOf( a.key ) !== -1
    const bSel = selected.indexOf( b.key ) !== -1

    if ( aSel && !bSel ) {
      return -1;
    }
    if ( !aSel && bSel ) {
      return 1;
    }

    // Both are selected or not selected
    // Sort by descending doc_count
    return b.doc_count - a.doc_count
  } )

  return retVal
}

/**
* Safely format a date
*
* @param {Date} date the date to convert
* @returns {string} the date formatted for the current locale
*/
export function shortFormat( date ) {
  const wrapped = moment( date )
  return date ? wrapped.format( 'M/D/YYYY' ) : ''
}

/**
* Convert a date to a truncated ISO-8601 string
*
* @param {Date} date the date to convert
* @returns {string} the date formatted as yyyy-mm-ddd
*/
export function shortIsoFormat( date ) {
  return date ? date.toISOString().substring( 0, 10 ) : ''
}

// ----------------------------------------------------------------------------
// attribution: underscore.js (MIT License)

/**
* Delay the implementation of a function until after a period of time
* This prevents expensive calls from being made while triggering events are
* still happening
*
* @param {function} func a function with an embedded expensive call
* @param {int} wait the number of msecs to delay before calling the function
* @returns {function} a replacement function to use in place of the original
*/
export function debounce( func, wait ) {
  var timer = null;

  var later = ( context, args ) => {
    timer = null;
    func.apply( context, args );
  }

  return ( ...theArgs ) => {
    if ( !timer ) {
      timer = setTimeout( later, wait, this, theArgs )
    }
  }
}

// ----------------------------------------------------------------------------
// attribution: lodash.js (Creative Commons License)

/**
* Binds methods of an object to the object itself, overwriting the existing
* method
*
* @param {Object} obj The object to bind and assign the bound methods to.
* @param {...(string|string[])} methodNames The object method names to bind,
*  specified individually or in arrays.
* @returns {Object} the updated object
*/
export function bindAll( obj, methodNames ) {
  const length = methodNames.length
  for ( let i = 0; i < length; i++ ) {
    const key = methodNames[i]
    obj[key] = obj[key].bind( obj )
  }
  return obj;
}
