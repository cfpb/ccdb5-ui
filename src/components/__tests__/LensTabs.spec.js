import configureMockStore from 'redux-mock-store';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import ReduxLensTabs, {
  LensTabs,
  mapDispatchToProps,
  mapStateToProps,
} from '../Trends/LensTabs';
import React from 'react';
import renderer from 'react-test-renderer';
import { REQUERY_ALWAYS } from '../../constants';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import * as utils from '../../utils';

/**
 *
 * @param root0
 * @param root0.focus
 * @param root0.lens
 * @param root0.results
 */
function setupSnapshot({ focus, lens, results }) {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  const store = mockStore({
    query: {
      focus,
      lens,
      subLens: 'sub_product',
    },
    trends: {
      results,
    },
  });

  return renderer.create(
    <Provider store={store}>
      <IntlProvider locale="en">
        <ReduxLensTabs showTitle={true} />
      </IntlProvider>
    </Provider>
  );
}

describe('component:LensTabs', () => {
  let fixture;
  beforeEach(() => {
    fixture = {
      focus: '',
      lens: 'Overview',
      results: {},
    };
  });

  it('does not render when Overview', () => {
    const target = setupSnapshot(fixture);
    const tree = target.toJSON();
    expect(tree).toBeNull();
  });

  it('renders Product without crashing', () => {
    fixture.lens = 'Product';
    const target = setupSnapshot(fixture);
    const tree = target.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders focus Product tab without crashing', () => {
    fixture.focus = 'fooBar';
    fixture.lens = 'Product';
    fixture.results = {
      'sub-product': [1, 2, 3],
    };

    const target = setupSnapshot(fixture);
    const tree = target.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('hides focus Product tab without crashing', () => {
    fixture.focus = 'fooBar';
    fixture.lens = 'Product';
    fixture.results = {
      issue: [1, 2, 3],
      'sub-product': [],
    };

    const target = setupSnapshot(fixture);
    const tree = target.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('buttons', () => {
    let cb = null;
    let target = null;

    beforeEach(() => {
      cb = jest.fn();
      target = shallow(
        <LensTabs
          onTab={cb}
          lens="Product"
          hasProductTab={true}
          subLens="Issue"
          showTitle={true}
        />
      );
    });

    it('tabChanged is called with Product when the button is clicked', () => {
      const prev = target.find('.tabbed-navigation button.sub_product');
      prev.simulate('click');
      expect(cb).toHaveBeenCalledWith('Product', 'sub_product');
    });

    it('tabChanged is called with Issue when the button is clicked', () => {
      const prev = target.find('.tabbed-navigation button.issue');
      prev.simulate('click');
      expect(cb).toHaveBeenCalledWith('Product', 'issue');
    });
  });

  describe('mapDispatchToProps', () => {
    it('hooks into changeDataSubLens', () => {
      const dispatch = jest.fn();
      const gaSpy = jest.spyOn(utils, 'sendAnalyticsEvent');
      mapDispatchToProps(dispatch).onTab('Some Lens', 'product');
      expect(dispatch.mock.calls).toEqual([
        [
          {
            meta: {
              requery: REQUERY_ALWAYS,
            },
            payload: 'product',
            type: 'trends/updateDataSubLens',
          },
        ],
      ]);
      expect(gaSpy).toHaveBeenCalledWith('Button', 'Some Lens:Products');
    });
  });

  describe('mapStateToProps', () => {
    it('maps state and props', () => {
      const state = {
        query: {
          focus: '',
          lens: 'foo',
          subLens: 'bar',
        },
        trends: {
          results: {
            foo: [],
          },
        },
      };
      const actual = mapStateToProps(state);
      expect(actual).toEqual({
        focus: '',
        lens: 'foo',
        hasProductTab: true,
        subLens: 'bar',
      });
    });
  });
});
