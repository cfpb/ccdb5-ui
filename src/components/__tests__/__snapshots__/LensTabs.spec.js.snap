// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component:LensTabs hides focus Product tab without crashing 1`] = `
<div
  className="tabbed-navigation lens"
>
  <section>
    <button
      className="tab issue"
      onClick={[Function]}
    >
      Issues
    </button>
  </section>
</div>
`;

exports[`component:LensTabs renders Product without crashing 1`] = `
<div
  className="tabbed-navigation lens"
>
  <section>
    <button
      className="tab sub_product active"
      onClick={[Function]}
    >
      Sub-products
    </button>
    <button
      className="tab issue"
      onClick={[Function]}
    >
      Issues
    </button>
  </section>
</div>
`;

exports[`component:LensTabs renders focus Product tab without crashing 1`] = `
<div
  className="tabbed-navigation lens"
>
  <section>
    <button
      className="tab sub_product active"
      onClick={[Function]}
    >
      Sub-products
    </button>
    <button
      className="tab issue"
      onClick={[Function]}
    >
      Issues
    </button>
  </section>
</div>
`;
